homeassistant/components/generic_thermostat/climate.py

```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
+                if self._is_cool_enabled:
+                    await self._async_ac_turn_off(force=force_resend)
+                if current_temp > target_temp_max:
+                    await self._async_heater_turn_off(force=force_resend)
+                elif current_temp <= target_temp_min:
+                    await self._async_heater_turn_on(force=force_resend)
+            elif self._hvac_mode == HVAC_MODE_COOL:
+                if self._is_heat_enabled:
+                    await self._async_heater_turn_off(force=force_resend)
+                if current_temp >= target_temp_max:
+                    await self._async_ac_turn_on(force=force_resend)
+                elif current_temp < target_temp_min:
+                    await self._async_ac_turn_off(force=force_resend)
+
+    async def _async_heater_turn_on(self, force=False):
+        """Turn heater toggleable device on."""
+        _LOGGER.debug("Turn Heater ON")
+        if self._is_heater_active and not force:
+            _LOGGER.debug("Heater already ON")
+            return
+        data = {ATTR_ENTITY_ID: self._heater_entity_id}
+        _LOGGER.debug("Order ON sent to heater device %s", self._heater_entity_id)
+        await self.hass.services.async_call(HA_DOMAIN, SERVICE_TURN_ON, data)
+
+    async def _async_heater_turn_off(self, force=False):
+        """Turn heater toggleable device off."""
+        _LOGGER.debug("Turn Heater OFF called")
+        if not self._is_heater_active and not force:
+            _LOGGER.debug("Heater already OFF")
+            return
+        data = {ATTR_ENTITY_ID: self._heater_entity_id}
+        _LOGGER.debug("Order OFF sent to heater device %s", self._heater_entity_id)
+        await self.hass.services.async_call(HA_DOMAIN, SERVICE_TURN_OFF, data)
+
+    async def _async_ac_turn_on(self, force=False):
+        """Turn ac toggleable device on."""
+        _LOGGER.debug("Turn AC ON")
+        if self._is_ac_active and not force:
+            _LOGGER.debug("AC already ON")
+            return
+        data = {ATTR_ENTITY_ID: self._ac_entity_id}
+        _LOGGER.debug("Order ON sent to AC device %s", self._ac_entity_id)
+        await self.hass.services.async_call(HA_DOMAIN, SERVICE_TURN_ON, data)
+
+    async def _async_ac_turn_off(self, force=False):
+        """Turn ac toggleable device off."""
+        _LOGGER.debug("Turn AC OFF")
+        if not self._is_ac_active and not force:
+            _LOGGER.debug("AC already OFF")
+            return
+        data = {ATTR_ENTITY_ID: self._ac_entity_id}
+        _LOGGER.debug("Order OFF sent to AC device %s", self._ac_entity_id)
+        await self.hass.services.async_call(HA_DOMAIN, SERVICE_TURN_OFF, data)
+
+    async def async_set_preset_mode(self, preset_mode: str):
+        """Set new preset mode.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        if preset_mode in self.preset_modes or preset_mode == PRESET_NONE:
```

Comment:
Invert this check and log error and return if true. Then we can outdent below.

Reply:
Done